xarray/testing/strategies.py

```diff
@@ -0,0 +1,482 @@
+from collections.abc import Hashable, Iterable, Mapping, Sequence
+from typing import TYPE_CHECKING, Any, Protocol, Union, overload
+
+try:
+    import hypothesis.strategies as st
+except ImportError as e:
+    raise ImportError(
+        "`xarray.testing.strategies` requires `hypothesis` to be installed."
+    ) from e
+
+import hypothesis.extra.numpy as npst
+import numpy as np
+from hypothesis.errors import InvalidArgument
+
+import xarray as xr
+from xarray.core.types import T_DuckArray
+
+if TYPE_CHECKING:
+    from xarray.core.types import _DTypeLikeNested, _ShapeLike
+
+
+__all__ = [
+    "supported_dtypes",
+    "names",
+    "dimension_names",
+    "dimension_sizes",
+    "attrs",
+    "variables",
+    "unique_subset_of",
+]
+
+
+class ArrayStrategyFn(Protocol[T_DuckArray]):
+    def __call__(
+        self,
+        *,
+        shape: "_ShapeLike",
+        dtype: "_DTypeLikeNested",
+    ) -> st.SearchStrategy[T_DuckArray]:
+        ...
+
+
+def supported_dtypes() -> st.SearchStrategy[np.dtype]:
+    """
+    Generates only those numpy dtypes which xarray can handle.
+
+    Use instead of hypothesis.extra.numpy.scalar_dtypes in order to exclude weirder dtypes such as unicode, byte_string, array, or nested dtypes.
+    Also excludes datetimes, which dodges bugs with pandas non-nanosecond datetime overflows.
+
+    Requires the hypothesis package to be installed.
+
+    See Also
+    --------
+    :ref:`testing.hypothesis`_
+    """
+    # TODO should this be exposed publicly?
+    # We should at least decide what the set of numpy dtypes that xarray officially supports is.
+    return (
+        npst.integer_dtypes()
+        | npst.unsigned_integer_dtypes()
+        | npst.floating_dtypes()
+        | npst.complex_number_dtypes()
+    )
+
+
+def smallish_arrays(
```

Comment:
That is the only reason. I did not think of using `functools.partial` - that's a good idea, I can try that out before merging.

Reply:
That actually won't work because we do need to be able to pass `shape` and `dtype` to the `array_strategy_fn`.

But I tried removing `smallish_arrays` completely and the tests still seem to complete in a reasonable amount of time, so I've actually just taken it out for now.